File: python/interview_37.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return '[]'
        res = []
        queue = collections.deque()
        nodeCount = 1
        queue.append(root)
        while nodeCount > 0:
            node = queue.popleft()
            if not node:
                res.append('null')
            else:
                nodeCount -= 1
                res.append(str(node.val))
                queue.append(node.left)
                if node.left:
                    nodeCount += 1
                queue.append(node.right)
                if node.right:
                    nodeCount += 1 
        return '[' + ','.join(res) + ']'

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if data == '[]':
            return None
        datas = data[1 : -1].split(',')
        n = len(datas)
        queue = collections.deque()
        root = TreeNode(int(datas[0]))
        queue.append(root)
        idx = 1
        while idx < n:
            node = queue.popleft()
            if datas[idx] != 'null':
                node.left = TreeNode(int(datas[idx]))
                queue.append(node.left)
            idx += 1
            if idx < n and datas[idx] != 'null':
                node.right = TreeNode(int(datas[idx]))
                queue.append(node.right)
            idx += 1
        return root

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: Java/interview_37.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) {
            return "[]";
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int nodeCount = 1;
        StringBuilder sb = new StringBuilder("[");
        while (nodeCount > 0) {
            TreeNode node = queue.poll();
            if (node == null) {
                sb.append("null,");
            }
            else {
                sb.append(node.val + ",");
                --nodeCount;
                queue.offer(node.left);
                if (node.left != null) {
                    ++nodeCount;
                }
                queue.offer(node.right);
                if (node.right != null) {
                    ++nodeCount;
                }
            }
        }
        sb.deleteCharAt(sb.length() - 1);
        sb.append("]");
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.equals("[]")) {
            return null;
        }
        String[] strs = data.substring(1, data.length() - 1).split(",");
        Queue<TreeNode> queue = new LinkedList<>();
        TreeNode root = new TreeNode(Integer.parseInt(strs[0]));
        queue.offer(root);
        int idx = 1;
        while (idx < strs.length) {
            TreeNode node = queue.poll();
            if (!strs[idx].equals("null")) {
                node.left = new TreeNode(Integer.parseInt(strs[idx]));
                queue.offer(node.left);
            }
            ++idx;
            if (idx < strs.length && !strs[idx].equals("null")) {
                node.right = new TreeNode(Integer.parseInt(strs[idx]));
                queue.offer(node.right);
            }
            ++idx;
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));
